yesod-core/Yesod/Core/Handler.hs

```diff
@@ -787,6 +789,30 @@ setHeader :: MonadHandler m => Text -> Text -> m ()
 setHeader = addHeader
 {-# DEPRECATED setHeader "Please use addHeader instead" #-}
 
+-- | Replace an existing header with a new value or add a new header
+-- if not present.
+--
+-- Note that, while the data type used here is 'Text', you must provide only
+-- ASCII value to be HTTP compliant.
+--
+-- @since 1.4.36
+replaceOrAddHeader :: MonadHandler m => Text -> Text -> m ()
+replaceOrAddHeader a b =
+  let header = Header (encodeUtf8 a) (encodeUtf8 b)
+  in modify $ \g -> g {ghsHeaders = replaceHeader header (ghsHeaders g)}
+  where
+    sameHeaderName :: Header -> Header -> Bool
+    sameHeaderName (Header n1 _) (Header n2 _) = n1 == n2
+    sameHeaderName _ _ = False
+
+    replaceHeader :: Header -> Endo [Header] -> Endo [Header]
+    replaceHeader header endo =
+      let allHeaders :: [Header] = appEndo endo []
+      in Endo
```

Comment:
@psibi I don't think so.  The behaviour of `addHeader` is that a later call puts the header later in the result, so to be consistent, and also "natural", I think, the result should have been (since Yesod appears to add `Vary` before the handler code):
```
Vary: Accept, Accept-Language
Foo: Baz
Foo: Dup
FOO: Final
```
Obviously that's with the case sensitive match still - actually we want just the Vary and the last FOO, of course!